modules/library/referencing/src/main/java/org/geotools/referencing/operation/matrix/GeneralMatrix.java

```diff
@@ -836,17 +911,41 @@ public final void mul(GeneralMatrix matrix){
      * @param matrix1
      * @param matrix2
      */
-    public void mul(GeneralMatrix matrix1, GeneralMatrix matrix2) {
-        if (mat.numRows == matrix1.mat.numRows && mat.numCols == matrix2.mat.numCols) {
-            DenseMatrix64F ret = new DenseMatrix64F(mat.getNumRows(), mat.getNumCols());
-            CommonOps.mult(matrix1.mat, matrix2.mat, ret);
-            mat = ret;
-        } else {
-            DenseMatrix64F ret = new DenseMatrix64F(matrix1.mat.numRows, matrix2.mat.numCols);
-            CommonOps.mult(matrix1.mat, matrix2.mat, ret);
-            mat = ret;
+    public void mul(Matrix matrix1, Matrix matrix2) {
+        DenseMatrix64F a = internal(matrix1);
+        DenseMatrix64F b = internal(matrix2);
+        if( a == mat || b == mat ){
+            mat = new DenseMatrix64F(a.numRows, b.numCols );
+        }
+        else {
+            mat.reshape(a.numRows, b.numCols, false);
         }
+        CommonOps.mult(a, b, mat);
+    }
+
```

Comment:
> Here I seem to see an optimization for nothing
> It seems the original intention was to leverage an in-place operation, to avoid instantiating another DenseMatrix64F, but then the implementation does create one regardless and replaces mat (so the original test has been done for nothing, or not?)

Agreed, was able to capture the sprit of the thing by checking if I can use reshape (to avoid creating a new DenseMatrix64F for the result.